python_modules/dagster/dagster/_core/definitions/multi_asset_sensor_definition.py

```diff
@@ -339,21 +340,21 @@ def _get_partitions_after_cursor(self, asset_key: AssetKey) -> List[str]:
             ]
         return partitions_to_fetch
 
-    def update_cursor_after_evaluation(self) -> None:
+    @property
+    def has_cursor_changed(self) -> bool:
+        return self._cursor_advance_state_mutation.has_advanced_records
+
+    def serialize_cursor(self) -> str:
         """Updates the cursor after the sensor evaluation function has been called. This method
         should be called at most once per evaluation.
         """
 
-        new_cursor = self._cursor_advance_state_mutation.get_cursor_with_advances(
-            self, self._unpacked_cursor
+        return json.dumps(
+            self._cursor_advance_state_mutation.get_cursor_with_advances(
+                self, self._unpacked_cursor
+            )
         )
 
-        if new_cursor != None:
-            # Cursor was not updated by this context object, so we do not need to update it
-            self._cursor = new_cursor
-            self._unpacked_cursor = MultiAssetSensorContextCursor(new_cursor, self)
```

Comment:
I think we should still keep these lines (354 and 355). When users create a context via `build_multi_asset_sensor_context`, the context should be reset after each sensor invocation so the sensor can be invoked multiple times with the same testing context

Reply:
Wrapping my head around this. Is there an example test that does this?

Reply:
Here's one: https://github.com/dagster-io/dagster/blob/f466eb77e35459165865b781137e7e744b1862f3/python_modules/dagster/dagster_tests/core_tests/definitions_tests/test_sensor_invocation.py#L768

Updating the cursor object at the end of the invocation means that you can fetch the updated state of the cursor after invoking and test against it. I suspect that re-adding this will resolve most of the bk errors